lib/words_report.rb

```diff
@@ -0,0 +1,28 @@
+class WordsReport
+  attr_reader :words
+
+  # Arbitrary definition of popular
+  POPULAR_COUNT = 5
+
+  def initialize(counts)
+    @counts = counts
+    @words = words_by_frequency
+  end
+
+  def popular_words(popular_count = POPULAR_COUNT)
+    @words.select { |word| @counts[word] >= popular_count }
+  end
+
+  def report(popular_count = POPULAR_COUNT)
+    popular_words(popular_count).each { |word| printf "%-12s: %3d\n", word, @counts[word] }
+  end
+
+  private
+
+  def words_by_frequency
+    @words = @counts.keys.sort do |a,b|
+      first = @counts[b] <=> @counts[a]
+      first == 0 ? a <=> b : first
```

Comment:
I try to avoid leaving data like this in an arbitrary order.
I have wasted lots of time in my career tracking down random problems that boiled down to data being reported in some random order. So, I tend to sort if the data is to be displayed or written to disk.

Now that I've changed it to the array compare, the code is more concise. I'll have to get used to the idiom for multi-key sorts.